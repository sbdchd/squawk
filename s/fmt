#!/bin/sh
set -eu

cargo fmt
