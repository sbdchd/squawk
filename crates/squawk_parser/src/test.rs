// based on https://github.com/rust-lang/rust-analyzer/blob/d8887c0758bbd2d5f752d5bd405d4491e90e7ed6/crates/parser/src/tests.rs
use camino::Utf8Path;
use dir_test::{dir_test, Fixture};
use insta::{assert_snapshot, with_settings};
use std::fmt::Write;

use crate::{parse, LexedStr};

fn parse_text(text: &str) -> (String, bool) {
    let lexed = LexedStr::new(text);
    let input = lexed.to_input();
    let output = parse(&input);

    let mut buf = String::new();
    let mut errors = Vec::new();
    let mut indent = String::new();
    let mut depth = 0;
    let mut len = 0;
    lexed.intersperse_trivia(&output, &mut |step| match step {
        crate::StrStep::Token { kind, text } => {
            assert!(depth > 0);
            len += text.len();
            writeln!(buf, "{indent}{kind:?} {text:?}").unwrap();
        }
        crate::StrStep::Enter { kind } => {
            assert!(depth > 0 || len == 0);
            depth += 1;
            writeln!(buf, "{indent}{kind:?}").unwrap();
            indent.push_str("  ");
        }
        crate::StrStep::Exit => {
            assert!(depth > 0);
            depth -= 1;
            indent.pop();
            indent.pop();
        }
        crate::StrStep::Error { msg, pos } => {
            assert!(depth > 0);
            let err = "ERROR";
            errors.push(format!("{err}@{pos}: {msg}\n"));
        }
    });
    assert_eq!(
        len,
        text.len(),
        "didn't parse all text.\nParsed:\n{}\n\nAll:\n{}\n",
        &text[..len],
        text
    );

    for (token, msg) in lexed.errors() {
        let pos = lexed.text_start(token);
        let err = "ERROR";
        errors.push(format!("{err}@{pos}: {msg}\n"));
    }

    let has_errors = !errors.is_empty();
    if has_errors {
        buf.push_str("---\n");
        for e in errors {
            buf.push_str(&e);
        }
    }
    (buf, has_errors)
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_data",
    glob: "**/*.sql",
)]
fn sqltest(fixture: Fixture<&str>) {
    let content = fixture.content();
    let absolute_fixture_path = Utf8Path::new(fixture.path());
    let input_file = absolute_fixture_path;
    let test_name = absolute_fixture_path
        .file_name()
        .and_then(|x| x.strip_suffix(".sql"))
        .unwrap();

    let parent_dir = input_file.parent().and_then(|x| x.file_name()).unwrap();

    let (parsed, has_errors) = parse_text(content);

    with_settings!({
      omit_expression => true,
      input_file => input_file,
    }, {
      assert_snapshot!(format!("{}_{}", test_name, parent_dir), parsed);
    });

    if parent_dir == "ok" {
        // We check that all of our tests in `ok` also pass the Postgres parser,
        // if they don't, they should be moved to the `err` directory.
        assert!(
            !has_errors,
            "tests defined in the `ok` can't have parser errors."
        );
        // skipping pg17 specific stuff since our parser isn't using the latest parser
        if !test_name.ends_with("pg17") {
            let pg_result = pg_query::parse(content);
            if let Err(e) = &pg_result {
                assert!(
                    &pg_result.is_ok(),
                    "tests defined in the `ok` can't have Postgres parser errors. Found {}",
                    e
                );
            }
        }
    } else {
        assert_eq!(parent_dir, "err");
        assert!(
            has_errors,
            "tests defined in the `err` directory must have parser errors."
        );
    }
}
