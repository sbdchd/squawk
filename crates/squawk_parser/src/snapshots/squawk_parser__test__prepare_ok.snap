---
source: crates/squawk_parser/src/test.rs
input_file: crates/squawk_parser/test_data/ok/prepare.sql
---
SOURCE_FILE
  COMMENT "-- insert"
  WHITESPACE "\n"
  PREPARE_STMT
    PREPARE_KW "PREPARE"
    WHITESPACE " "
    NAME
      IDENT "fooplan"
    WHITESPACE " "
    L_PAREN "("
    PATH_TYPE
      PATH
        PATH_SEGMENT
          NAME_REF
            INT_KW "int"
    COMMA ","
    WHITESPACE " "
    PATH_TYPE
      PATH
        PATH_SEGMENT
          NAME_REF
            TEXT_KW "text"
    COMMA ","
    WHITESPACE " "
    PATH_TYPE
      PATH
        PATH_SEGMENT
          NAME_REF
            IDENT "bool"
    COMMA ","
    WHITESPACE " "
    PATH_TYPE
      PATH
        PATH_SEGMENT
          NAME_REF
            NUMERIC_KW "numeric"
    R_PAREN ")"
    WHITESPACE " "
    AS_KW "AS"
    WHITESPACE "\n    "
    INSERT_STMT
      INSERT_KW "INSERT"
      WHITESPACE " "
      INTO_KW "INTO"
      WHITESPACE " "
      PATH
        PATH_SEGMENT
          NAME_REF
            IDENT "foo"
      WHITESPACE " "
      SELECT
        VALUES_KW "VALUES"
        L_PAREN "("
        LITERAL
          PARAM "$1"
        COMMA ","
        WHITESPACE " "
        LITERAL
          PARAM "$2"
        COMMA ","
        WHITESPACE " "
        LITERAL
          PARAM "$3"
        COMMA ","
        WHITESPACE " "
        LITERAL
          PARAM "$4"
        R_PAREN ")"
  SEMICOLON ";"
  WHITESPACE "\n\n\n\n"
  COMMENT "-- select"
  WHITESPACE "\n"
  PREPARE_STMT
    PREPARE_KW "PREPARE"
    WHITESPACE " "
    NAME
      IDENT "usrrptplan"
    WHITESPACE " "
    L_PAREN "("
    PATH_TYPE
      PATH
        PATH_SEGMENT
          NAME_REF
            INT_KW "int"
    R_PAREN ")"
    WHITESPACE " "
    AS_KW "AS"
    WHITESPACE "\n    "
    SELECT
      SELECT_CLAUSE
        SELECT_KW "SELECT"
        WHITESPACE " "
        TARGET_LIST
          TARGET
            STAR "*"
      WHITESPACE " "
      FROM_CLAUSE
        FROM_KW "FROM"
        WHITESPACE " "
        NAME_REF
          IDENT "users"
        WHITESPACE " "
        ALIAS
          NAME
            IDENT "u"
        COMMA ","
        WHITESPACE " "
        NAME_REF
          IDENT "logs"
        WHITESPACE " "
        ALIAS
          NAME
            IDENT "l"
      WHITESPACE " "
      WHERE_CLAUSE
        WHERE_KW "WHERE"
        WHITESPACE " "
        BIN_EXPR
          BIN_EXPR
            BIN_EXPR
              FIELD_EXPR
                NAME_REF
                  IDENT "u"
                DOT "."
                NAME_REF
                  IDENT "usrid"
              EQ "="
              LITERAL
                PARAM "$1"
            WHITESPACE " "
            AND_KW "AND"
            WHITESPACE " "
            BIN_EXPR
              FIELD_EXPR
                NAME_REF
                  IDENT "u"
                DOT "."
                NAME_REF
                  IDENT "usrid"
              EQ "="
              FIELD_EXPR
                NAME_REF
                  IDENT "l"
                DOT "."
                NAME_REF
                  IDENT "usrid"
          WHITESPACE "\n    "
          AND_KW "AND"
          WHITESPACE " "
          BIN_EXPR
            FIELD_EXPR
              NAME_REF
                IDENT "l"
              DOT "."
              NAME_REF
                IDENT "date"
            WHITESPACE " "
            EQ "="
            WHITESPACE " "
            LITERAL
              PARAM "$2"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  PREPARE_STMT
    PREPARE_KW "PREPARE"
    WHITESPACE " "
    NAME
      IDENT "foo"
    WHITESPACE " "
    AS_KW "AS"
    WHITESPACE "\n    "
    SELECT
      WITH_CLAUSE
        WITH_KW "with"
        WHITESPACE " "
        WITH_TABLE
          NAME
            IDENT "t"
          WHITESPACE " "
          AS_KW "as"
          WHITESPACE " "
          L_PAREN "("
          SELECT
            SELECT_CLAUSE
              SELECT_KW "select"
              WHITESPACE " "
              TARGET_LIST
                TARGET
                  LITERAL
                    INT_NUMBER "1"
          R_PAREN ")"
      WHITESPACE "\n    "
      SELECT_CLAUSE
        SELECT_KW "select"
        WHITESPACE " "
        TARGET_LIST
          TARGET
            STAR "*"
      WHITESPACE " "
      FROM_CLAUSE
        FROM_KW "from"
        WHITESPACE " "
        NAME_REF
          IDENT "t"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- update"
  WHITESPACE "\n"
  PREPARE_STMT
    PREPARE_KW "prepare"
    WHITESPACE " "
    NAME
      IDENT "foo"
    WHITESPACE " "
    AS_KW "as"
    WHITESPACE "\n  "
    UPDATE_STMT
      UPDATE_KW "update"
      WHITESPACE " "
      PATH
        PATH_SEGMENT
          NAME_REF
            IDENT "foo"
      WHITESPACE " "
      SET_KW "set"
      WHITESPACE " "
      NAME_REF
        IDENT "x"
      WHITESPACE " "
      EQ "="
      WHITESPACE " "
      LITERAL
        INT_NUMBER "1"
      WHITESPACE " "
      WHERE_CLAUSE
        WHERE_KW "where"
        WHITESPACE " "
        BIN_EXPR
          NAME_REF
            IDENT "x"
          WHITESPACE " "
          R_ANGLE ">"
          WHITESPACE " "
          LITERAL
            INT_NUMBER "10"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- delete"
  WHITESPACE "\n"
  PREPARE_STMT
    PREPARE_KW "prepare"
    WHITESPACE " "
    NAME
      IDENT "foo"
    WHITESPACE " "
    AS_KW "as"
    WHITESPACE "\n  "
    DELETE_STMT
      DELETE_KW "delete"
      WHITESPACE " "
      FROM_KW "from"
      WHITESPACE " "
      PATH
        PATH_SEGMENT
          NAME_REF
            IDENT "foo"
      WHITESPACE " "
      WHERE_CLAUSE
        WHERE_KW "where"
        WHITESPACE " "
        BIN_EXPR
          NAME_REF
            IDENT "x"
          WHITESPACE " "
          EQ "="
          WHITESPACE " "
          LITERAL
            INT_NUMBER "1"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- merge"
  WHITESPACE "\n"
  PREPARE_STMT
    PREPARE_KW "prepare"
    WHITESPACE " "
    NAME
      IDENT "foo"
    WHITESPACE " "
    AS_KW "as"
    WHITESPACE "\n  "
    MERGE_STMT
      MERGE_KW "merge"
      WHITESPACE " "
      INTO_KW "into"
      WHITESPACE " "
      PATH
        PATH_SEGMENT
          NAME_REF
            IDENT "t1"
      WHITESPACE "\n  "
      USING_CLAUSE
        USING_KW "using"
        WHITESPACE " "
        NAME_REF
          IDENT "t2"
        WHITESPACE " "
        ON_KW "on"
        WHITESPACE " "
        BIN_EXPR
          FIELD_EXPR
            NAME_REF
              IDENT "t2"
            DOT "."
            NAME_REF
              IDENT "id"
          WHITESPACE " "
          EQ "="
          WHITESPACE " "
          FIELD_EXPR
            NAME_REF
              IDENT "t1"
            DOT "."
            NAME_REF
              IDENT "id"
      WHITESPACE "\n  "
      WHEN_KW "when"
      WHITESPACE " "
      MATCHED_KW "matched"
      WHITESPACE " "
      THEN_KW "then"
      WHITESPACE " "
      DO_KW "do"
      WHITESPACE " "
      NOTHING_KW "nothing"
      WHITESPACE "\n  "
      WHEN_KW "when"
      WHITESPACE " "
      NOT_KW "not"
      WHITESPACE " "
      MATCHED_KW "matched"
      WHITESPACE " "
      BY_KW "by"
      WHITESPACE " "
      SOURCE_KW "source"
      WHITESPACE " "
      THEN_KW "then"
      WHITESPACE " "
      DO_KW "do"
      WHITESPACE " "
      NOTHING_KW "nothing"
      WHITESPACE "\n  "
      WHEN_KW "when"
      WHITESPACE " "
      NOT_KW "not"
      WHITESPACE " "
      MATCHED_KW "matched"
      WHITESPACE " "
      THEN_KW "then"
      WHITESPACE " "
      DO_KW "do"
      WHITESPACE " "
      NOTHING_KW "nothing"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- values"
  WHITESPACE "\n"
  PREPARE_STMT
    PREPARE_KW "prepare"
    WHITESPACE " "
    NAME
      IDENT "foo"
    WHITESPACE " "
    AS_KW "as"
    WHITESPACE "\n  "
    SELECT
      VALUES_KW "values"
      WHITESPACE " "
      L_PAREN "("
      LITERAL
        INT_NUMBER "1"
      COMMA ","
      WHITESPACE " "
      LITERAL
        STRING "'one'"
      R_PAREN ")"
      COMMA ","
      WHITESPACE " "
      L_PAREN "("
      LITERAL
        INT_NUMBER "2"
      COMMA ","
      WHITESPACE " "
      LITERAL
        STRING "'two'"
      R_PAREN ")"
  SEMICOLON ";"
  WHITESPACE "\n\n"
